File: 프로그래머스/3/150367. 표현 가능한 이진트리/표현 가능한 이진트리.cpp
#include <bits/stdc++.h>
using namespace std;

string hexToDec(long long num){
    string ret = "";
    
    while(num){
        ret = to_string(num % 2) + ret;
        num /= 2;
    }
    return ret;
}

string fullDec(string s){
    string ret = s;
    int idx = 2;
    while(true){
        if(s.length() < idx) break;
        idx *= 2;
    }
    for(int i = 0; i < idx - 1 - s.length(); i++) ret = "0" + ret;
    return ret;
}

bool possible(string s){
    if(s == "1") return true;
    
    bool allZero = true;
    for(int i = 0; i<s.length(); i++){
        if(s[i] == '1') allZero = false;
    }
    
    if(allZero) return true;
    
    int mid = s.length()/2;
    string left = s.substr(0,mid);
    string right = s.substr(mid + 1);
    if(s[mid] == '1' && possible(left) && possible(right)) return true;
    return false;
}

vector<int> solution(vector<long long> numbers) {
    vector<int> answer;
    for(const auto num : numbers){
        string dec = hexToDec(num);
        string fDec = fullDec(dec);
        answer.push_back(possible(fDec));
    }
    return answer;
}

File: 프로그래머스/3/150367. 표현 가능한 이진트리/표현 가능한 이진트리.java
import java.util.*;

class Solution {
    public int[] solution(long[] numbers) {
        int[] answer = new int[numbers.length];
        
        for(int i = 0 ; i< numbers.length; i++){
            String bin = Long.toBinaryString(numbers[i]);
            int len = 1;
            
            while(len < bin.length()){
                len = len * 2 + 1;
            }
            
            
            String tmp = "";
            for(int j = 0; j<len - bin.length(); j++){
                tmp += "0";
            }
            
            tmp += bin;
            bin = tmp;
            
            answer[i] = check(bin) ? 1 : 0;
        }
        
        return answer;
    }
    
    boolean check(String bin){
        int n = bin.length();
        if(n == 1) return true;
        
        char root = bin.charAt(n/2);
        String left = bin.substring(0, n / 2);
        String right = bin.substring(n / 2 + 1);
        
        if(root == '0' && (left.contains("1") || right.contains("1"))) return false;
        
        return check(left) && check(right);
    }

}